src/github.com/matrix-org/dendrite/mediaapi/writers/download.go

```diff
@@ -101,3 +118,152 @@ func (r *downloadRequest) Validate() *util.JSONResponse {
 	}
 	return nil
 }
+
+func (r *downloadRequest) doDownload(w http.ResponseWriter, cfg *config.MediaAPI, db *storage.Database, activeRemoteRequests *types.ActiveRemoteRequests) *util.JSONResponse {
+	// check if we have a record of the media in our database
+	mediaMetadata, err := r.getMediaMetadata(db)
+	if err == nil {
+		// If we have a record, we can respond from the local file
+		r.MediaMetadata = mediaMetadata
+		return r.respondFromLocalFile(w, cfg.AbsBasePath)
+	} else if err == errDBNotFound {
+		if r.MediaMetadata.Origin == cfg.ServerName {
+			// If we do not have a record and the origin is local, the file is not found
+			r.Logger.WithError(err).Warn("Failed to look up file in database")
+			return &util.JSONResponse{
+				Code: 404,
+				JSON: jsonerror.NotFound(fmt.Sprintf("File with media ID %q does not exist", r.MediaMetadata.MediaID)),
+			}
+		}
+		// TODO: If we do not have a record and the origin is remote, we need to fetch it and respond with that file
+		return &util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound(fmt.Sprintf("File with media ID %q does not exist", r.MediaMetadata.MediaID)),
+		}
+	}
+	// Another error from the database
+	r.Logger.WithError(err).WithFields(log.Fields{
+		"MediaID": r.MediaMetadata.MediaID,
+		"Origin":  r.MediaMetadata.Origin,
+	}).Error("Error querying the database.")
+	return &util.JSONResponse{
+		Code: 500,
+		JSON: jsonerror.InternalServerError(),
+	}
+}
+
+// getMediaMetadata queries the database for media metadata
+func (r *downloadRequest) getMediaMetadata(db *storage.Database) (*types.MediaMetadata, error) {
+	mediaMetadata, err := db.GetMediaMetadata(r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			r.Logger.WithFields(log.Fields{
+				"Origin":  r.MediaMetadata.Origin,
+				"MediaID": r.MediaMetadata.MediaID,
+			}).Info("Media not found in database.")
+			return nil, errDBNotFound
+		}
+		r.Logger.WithError(err).WithFields(log.Fields{
+			"Origin":  r.MediaMetadata.Origin,
+			"MediaID": r.MediaMetadata.MediaID,
+		}).Error("Error querying database for media.")
+		return nil, errDBQuery
+	}
+	return mediaMetadata, nil
+}
+
+// respondFromLocalFile reads a file from local storage and writes it to the http.ResponseWriter
+// Returns a util.JSONResponse error in case of error
+func (r *downloadRequest) respondFromLocalFile(w http.ResponseWriter, absBasePath types.Path) *util.JSONResponse {
+	filePath, err := fileutils.GetPathFromBase64Hash(r.MediaMetadata.Base64Hash, absBasePath)
+	if err != nil {
+		// FIXME: Remove erroneous file from database?
+		r.Logger.WithError(err).Warn("Failed to get file path from metadata")
+		return &util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound(fmt.Sprintf("File with media ID %q does not exist", r.MediaMetadata.MediaID)),
+		}
+	}
+	file, err := os.Open(filePath)
+	defer file.Close()
+	if err != nil {
+		// FIXME: Remove erroneous file from database?
```

Comment:
I'd probably just return a 500 response if the filesytem and database have become out of sync like this.